iceberg/streams/iceberg/file_reader_builder.cpp

```diff
@@ -0,0 +1,126 @@
+#include "iceberg/streams/iceberg/file_reader_builder.h"
+
+namespace iceberg {
+
+namespace {
+
+void InsertOrFail(std::unordered_set<int>& set, int x) {
+  Ensure(!set.contains(x), "Field ids are not unique");
+  set.insert(x);
+}
+
+std::unordered_map<int, int> MakeIndexToFieldIdMap(const parquet::schema::GroupNode* node,
+                                                   const std::optional<std::string>& schema_name_mapping) {
+  if (!schema_name_mapping.has_value()) {
+    return {};
+  }
+  std::unordered_set<int> field_ids_set;
+  std::optional<SchemaNameMapper> schema_name_mapper;
+  std::optional<SchemaNameMapper::Node> root_node;
+  std::unordered_map<int, int> index_to_field_id;
+  for (int i = 0; i < node->field_count(); ++i) {
+    if (node->field(i)->field_id() >= 0) {
+      InsertOrFail(field_ids_set, node->field(i)->field_id());
+      continue;
+    }
+    if (!schema_name_mapper.has_value()) {
+      schema_name_mapper = SchemaNameMapper(*schema_name_mapping);
+      root_node = schema_name_mapper.value().GetRootNode();
+    }
+    auto field_name = node->field(i)->name();
+    auto field_id = root_node->GetFieldIdByName(field_name);
```

Comment:
There's a bad naming here. There're 2 nodes: parquet one and json one. And you add another `SchemaNameMapper::Node`.
In this context there should be only one entity that could be called as `node` - a parquet node. There's no reason to tell about json nodes here. They should be hidden by SchemaNameMapper's iface. You make SchemaNameMapper to hide json from class user. And it's possible to rename `SchemaNameMapper::Node` to something unique.